### ISSUE TITLE
CA Rotation with Windows Desktop clients fails.

### ISSUE THREAD

**Issue by Valien** [gravitational/teleport#12636]
Expected behavior:

Following our [guides](https://goteleport.com/docs/setup/operations/ca-rotation/#certificate-authority-rotation) on CA Rotation (host) it should work on both manual and auto. 

Current behavior:

If you have a Windows Desktop service enabled in your environment then it will fail/get stuck during the process. You will need to disable the Windows Desktop service in order for the rotation to proceed.

Bug details:
- Teleport version: `v9.1.2` (both server, and teleport cloud). `v9.2` for local mac `tsh` client.
- Recreation steps...

Server to note: `av-teleport-demo-server-2` -- this is the Linux node that is running Desktop Access protocol

Verify your `teleport.yaml` file:

````
windows_desktop_service:
  enabled: yes
````

Run a `tctl status` to get current cluster info:

````
tctl status
Cluster  <CLUSTER>.teleport.sh
Version  9.1.2
Host CA  rotated May 12 20:16:52 UTC
User CA  rotated May 12 14:03:14 UTC
Jwt CA   rotated May 12 14:03:14 UTC
CA pin   sha256:XXX571
````

Get status of nodes:

````
tctl get nodes --format=json | jq '.[] | {hostname: .spec.hostname, rotation: .spec.rotation.state, phase: .spec.rotation.phase}'
````

````
{
  "hostname": "teleport-demo-app-server",
  "rotation": "standby",
  "phase": "standby"
}
{
  "hostname": "av-teleport-demo-server-2",
  "rotation": "standby",
  "phase": "standby"
}
{
  "hostname": "teleport-demo-k8s-server",
  "rotation": "standby",
  "phase": "standby"
}
````

Start the rotation: `tctl auth rotate --phase=init --manual --type=host`

View status:

````
tctl status
Cluster  <CLUSTER>.teleport.sh                                                            
Version  9.1.2                                                                          
Host CA  initialized (mode: manual, started: May 13 13:21:47 UTC, ending: May 14 19:21:47 UTC)
User CA  rotated May 12 14:03:14 UTC                                                    
Jwt CA   rotated May 12 14:03:14 UTC                                                    
CA pin   sha256:XXX571        
CA pin   sha256:XXXccb        
````

View node status:

````
{
  "hostname": "teleport-demo-app-server",
  "rotation": "in_progress",
  "phase": "init"
}
{
  "hostname": "av-teleport-demo-server-2",
  "rotation": "standby",
  "phase": "standby"
}
{
  "hostname": "teleport-demo-k8s-server",
  "rotation": "in_progress",
  "phase": "init"
}
````

Notice that all nodes go to `init` but the server with Desktop Access is stuck on `standby`. I waited upwards of 15-20 minutes and its status never changed.

Go into the node and set the following:

````
windows_desktop_service:
  enabled: no
````

Restart the Teleport service and view the nodes:

````
{
  "hostname": "teleport-demo-app-server",
  "rotation": "in_progress",
  "phase": "init"
}
{
  "hostname": "av-teleport-demo-server-2",
  "rotation": "in_progress",
  "phase": "init"
}
{
  "hostname": "teleport-demo-k8s-server",
  "rotation": "in_progress",
  "phase": "init"
}
````

Within a minute or so the node that was in `standby` goes into `init` phase.

Proceed with the rotation with `tctl auth rotate --phase=update_clients --manual` and all seems to be working ok. 

````
{
  "hostname": "teleport-demo-app-server",
  "rotation": "in_progress",
  "phase": "update_clients"
}
{
  "hostname": "av-teleport-demo-server-2",
  "rotation": "in_progress",
  "phase": "update_clients"
}
{
  "hostname": "teleport-demo-k8s-server",
  "rotation": "in_progress",
  "phase": "update_clients"
}
````

If you are on the node and viewing the logs with `journalctl -fu teleport` you'll see activity as it gets the CA rotation info and updates it's local information.

One sidenote (not listed in the docs) is that you might get a local error like the following:

`tctl auth rotate --phase=update_clients --manual
ERROR: can only switch to phase update_clients from init, current phase is standby`

But it still proceeds to progress to the next phase without issue.

Once all nodes are updated go to the next phase:

````
tctl auth rotate --phase=update_servers --manual
ERROR: can only switch to phase update_servers from update_clients, current phase is standby
````

Still getting this error but it still proceeds. But you do get this (at least in my environment):

````
[CLIENT]       Cannot connect to the auth server: failed direct dial to auth server: Get "https://teleport.cluster.local/v2/configuration/name": x509: certificate is valid for *.<CLUSTER>.teleport.sh, <CLUSTER>.teleport.sh, not 736b79776172642e74656c65706f72742e7368.teleport.cluster.local
	Get "https://teleport.cluster.local/v2/configuration/name": x509: certificate is valid for *.<CLUSTER>.teleport.sh, <CLUSTER>.teleport.sh, not 736b79776172642e74656c65706f72742e7368.teleport.cluster.local, failed dial to auth server through reverse tunnel: Get "https://teleport.cluster.local/v2/configuration/name": ssh: handshake failed: ssh: no authorities for hostname: <CLUSTER>.teleport.sh:443
	Get "https://teleport.cluster.local/v2/configuration/name": ssh: handshake failed: ssh: no authorities for hostname: <CLUSTER>.teleport.sh:443.
Is the auth server running on "<CLUSTER>.teleport.sh:443"?
````

Locally all I had to do is re-run a new `tsh login` and it re-logged me into Teleport and then was able to run my `tctl` commands again.

Updated status:

````
tctl status
Cluster  <CLUSTER>.teleport.sh                                                            
Version  9.1.2                                                                          
Host CA  rotating servers (mode: manual, started: May 13 13:21:47 UTC, ending: May 14 19:21:47 UTC)
User CA  rotated May 12 14:03:14 UTC                                                    
Jwt CA   rotated May 12 14:03:14 UTC                                                    
CA pin   sha256:4c9e3a41730c082dc2a18b9ae139f4a807571936fce4d6ea397abaf1e8a47ccb        
CA pin   sha256:8e9c2d42f87d7541bffbc70a3c0fbff73a061ca06ee8742023c476c81546d571        
````
Still good here!

````
{
  "hostname": "teleport-demo-app-server",
  "rotation": "in_progress",
  "phase": "update_servers"
}
{
  "hostname": "av-teleport-demo-server-2",
  "rotation": "in_progress",
  "phase": "update_servers"
}
{
  "hostname": "teleport-demo-k8s-server",
  "rotation": "in_progress",
  "phase": "update_servers"
}
````

All is good here as well!

Verify connectivity to the nodes via browser, and ssh and all works. No problems `tsh ssh` into any nodes.

Migrate to the final phase:

````
tctl auth rotate --phase=standby --manual
ERROR: can not transition to phase "standby" from "standby" phase.
````

Status works...

````
tctl status
Cluster  <CLUSTER>.teleport.sh
Version  9.1.2
Host CA  rotated May 13 14:16:25 UTC
User CA  rotated May 12 14:03:14 UTC
Jwt CA   rotated May 12 14:03:14 UTC
CA pin   sha256:4c9e3a41730c082dc2a18b9ae139f4a807571936fce4d6ea397abaf1e8a47ccb
````

And nodes status:

````
{
  "hostname": "teleport-demo-app-server",
  "rotation": "standby",
  "phase": "standby"
}
{
  "hostname": "av-teleport-demo-server-2",
  "rotation": "standby",
  "phase": "standby"
}
{
  "hostname": "teleport-demo-k8s-server",
  "rotation": "standby",
  "phase": "standby"
}
````

Per the guide, if you rotated the `User CA` then you will need to upload and add the new `User CA` to the Windows server. 

Now that the rotation is complete, re-enable the Desktop Service in the `teleport.yaml` and restart the service. You will not be able to connect to the Desktop service. It might show up in the Teleport UI but on connecting, it will not connect. 

Error logs from the linux node that has the desktop service:

````
May 13 14:53:00 ip-172-16-10-161 teleport[7888]: 2022-05-13T14:53:00Z ERRO [PROC:1]    "WindowsDesktop failed to establish connection to cluster: Failed to connect to Auth Server directly or over tunnel, no methods remaining.\n\tGet \"https://teleport.cluster.local/v2/domain\": x509: certificate is valid for *.<CLUSTER>.teleport.sh, <CLUSTER>.teleport.sh, not 736b79776172642e74656c65706f72742e7368.teleport.cluster.local, Get \"https://teleport.cluster.local/v2/domain\": ssh: handshake failed: ssh: no authorities for hostname: <CLUSTER>.teleport.sh:443." service/connect.go:86
````

If you stop the Teleport service, remove the `/var/lib/teleport` directory and then start Teleport you'll get a failure with an invalid join token. So now you have to create a new join token and add it. Once the node connects to the cluster it will show up in the UI and the Windows desktop will also show up in the UI. You can then attempt to connect and will get a spinning circle and the following logs will show this:

````
May 13 14:59:00 ip-172-16-10-161 teleport[8166]: 2022-05-13T14:59:00Z WARN [WINDOWS_D] mTLS authentication failed for incoming connection client-ip:35.71.146.110 error:[EOF] desktop/windows_server.go:708
May 13 14:59:00 ip-172-16-10-161 teleport[8166]: 2022-05-13T14:59:00Z ERRO [WINDOWS_D] Failed to send TDP error message EOF desktop/windows_server.go:677
May 13 15:02:21 ip-172-16-10-161 teleport[8166]: 2022-05-13T15:02:21Z INFO [WINDOWS_D] windows_desktop teleportwin-teleportdemo-local updated, updating. services/reconciler.go:169
May 13 15:07:21 ip-172-16-10-161 teleport[8166]: 2022-05-13T15:07:21Z INFO [WINDOWS_D] windows_desktop teleportwin-teleportdemo-local updated, updating. services/reconciler.go:169
````

TL;DR:

Rotating CA certs with Desktop Access will fail. Disabling desktop access will let the rotation continue but the Desktop will never be able to be accessed (could be a different issue) even when updating a new `user ca` cert, etc.

Only thing I haven't done is totally remove Teleport settings off Windows Server as well as the node itself. But that's not a good experience for customers.

- Debug logs

**Comment by zmb3** [comment]
See also #10111, which is the short term fix before we support proper CA rotation for desktop access.

**Comment by zmb3** [comment]
This was fixed by #14804.
